sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp

```diff
@@ -67,524 +126,460 @@ static sycl::detail::ESIMDEmuPluginOpaqueData *PiESimdDeviceAccess;
-  // TODO : construct cl::sycl::nd_item<1> instance using thread ID
-  // info retrieved from CM with GlobalOffset/GlobalSize/LocalSize
-  // info and call Lambda function
-  // LambdaWrapper->Func(nd_item_1dim);
-}
+  const sycl::id<NDims> GroupID = RangeBuilder<NDims>::create(
+      [](int i) { return cm_support::get_group_idx(i); });
 
-extern "C" inline void invokeLambda_NDITEM_2DIM(void *Wrapper) {
-  auto *LambdaWrapper = reinterpret_cast<LambdaWrapper_NDITEM_2DIM *>(Wrapper);
+  const sycl::group<NDims> Group = IDBuilder::createGroup<NDims>(
+      LambdaWrapper->GlobalSize, LambdaWrapper->LocalSize, GroupSize, GroupID);
 
-  PLACEHOLDER_UNUSED(LambdaWrapper);
-  // TODO : construct cl::sycl::nd_item<2> instance using thread ID
-  // info retrieved from CM with GlobalOffset/GlobalSize/LocalSize
-  // info and call Lambda function
-  // LambdaWrapper->Func(nd_item_2dim);
-}
+  const sycl::id<NDims> GlobalID = GroupID * LambdaWrapper->LocalSize +
+                                   LocalID + LambdaWrapper->GlobalOffset;
+  const sycl::item<NDims, /*Offset=*/true> GlobalItem =
+      IDBuilder::createItem<NDims, true>(LambdaWrapper->GlobalSize, GlobalID,
+                                         LambdaWrapper->GlobalOffset);
+  const sycl::item<NDims, /*Offset=*/false> LocalItem =
+      IDBuilder::createItem<NDims, false>(LambdaWrapper->LocalSize, LocalID);
 
-extern "C" inline void invokeLambda_NDITEM_3DIM(void *Wrapper) {
-  auto *LambdaWrapper = reinterpret_cast<LambdaWrapper_NDITEM_3DIM *>(Wrapper);
+  const sycl::nd_item<NDims> NDItem =
+      IDBuilder::createNDItem<NDims>(GlobalItem, LocalItem, Group);
 
-  PLACEHOLDER_UNUSED(LambdaWrapper);
-  // TODO : construct cl::sycl::nd_item<3> instance using thread ID
-  // info retrieved from CM with GlobalOffset/GlobalSize/LocalSize
-  // info and call Lambda function
-  // LambdaWrapper->Func(nd_item_3dim);
+  LambdaWrapper->Func(NDItem);
 }
 
 // libCMBatch class defines interface for lauching kernels with
 // software multi-threads
-template <class KernelType, class KernelArgType, int DIMS> class libCMBatch {
+template <int DIMS> class libCMBatch {
 private:
   // Kernel function
-  KernelType MKernel;
+  KernelFunc<DIMS> MKernel;
 
   // Space-dimension info
   std::vector<uint32_t> GroupDim;
   std::vector<uint32_t> SpaceDim;
 
-  // Number of threads for parallelization
-  const uint32_t hwThreads = (uint32_t)std::thread::hardware_concurrency();
-
-  using IDBuilder = sycl::detail::Builder;
-  const sycl::id<DIMS> UnusedID =
-      sycl::detail::InitializedVal<DIMS, sycl::id>::template get<0>();
-  const sycl::range<DIMS> UnusedRange =
-      sycl::detail::InitializedVal<DIMS, sycl::range>::template get<0>();
-
 public:
-  libCMBatch(KernelType Kernel)
+  libCMBatch(KernelFunc<DIMS> Kernel)
       : MKernel(Kernel), GroupDim{1, 1, 1}, SpaceDim{1, 1, 1} {
-    assert(MKernel != nullptr);
+    if (MKernel == nullptr) {
+      throw sycl::kernel_error();
```

Comment:
Usually plugins do not let exceptions leave their interfaces, suggest using error codes instead.

Reply:
`libCMBatch` here is class constructor. As far as I know, exception throwing is recommended for constructor failure.

Reply:
Sanity check for kernel pointer(`MKernel` here) is moved to `piEnqueueKernelLaunch`.